lib/src/migrator.dart

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018 Google LLC
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+// The sass package's API is not necessarily stable. It is being imported with
+// the Sass team's explicit knowledge and approval. See
+// https://github.com/sass/dart-sass/issues/236.
+import 'package:sass/src/ast/sass.dart';
+import 'package:sass/src/visitor/recursive_statement.dart';
+import 'package:sass/src/visitor/interface/expression.dart';
+
+import 'package:path/path.dart' as p;
+
+import 'built_in_functions.dart';
+import 'local_scope.dart';
+import 'patch.dart';
+import 'stylesheet_migration.dart';
+import 'utils.dart';
+
+/// Runs a migration of [entrypoints] and their dependencies without writing
+/// any changes to disk.
+p.PathMap<String> migrateFiles(List<String> entrypoints) =>
+    _Migrator().migrate(entrypoints);
+
+class _Migrator extends RecursiveStatementVisitor implements ExpressionVisitor {
+  /// List of all migrations for files touched by this run.
+  p.PathMap<StylesheetMigration> _migrations = p.PathMap();
+
+  /// List of migrations in progress. The last item in the current migration.
+  List<StylesheetMigration> _activeMigrations = [];
+
+  StylesheetMigration get _currentMigration =>
+      _activeMigrations.isNotEmpty ? _activeMigrations.last : null;
+
+  /// Runs the migrator on each item in [entrypoints] in turn and returns a
+  /// map of migrated contents.
+  p.PathMap<String> migrate(List<String> entrypoints) {
+    for (var entrypoint in entrypoints) {
+      _migrateStylesheet(entrypoint);
+    }
+    var results = p.PathMap<String>();
+    for (var migration in _migrations.values) {
+      results[migration.path] = migration.migratedContents;
+    }
+    return results;
+  }
+
+  /// Migrates the stylesheet at [path] if it hasn't already been migrated and
+  /// returns the StylesheetMigration instance for it regardless.
+  StylesheetMigration _migrateStylesheet(String path) {
+    path = canonicalizePath(path,
+        loadPath: _currentMigration == null
+            ? p.current
+            : p.dirname(_currentMigration.path));
```

Comment:
Setting the load path to the directory of the current file is probably right. I'm guessing you're doing that to resolve relative imports, but they should instead be handled by interpreting `path` as relative to the entrypoint's path. So, you'd call:

```dart
path = canonicalizePath(
    _currentMigration == null ? path : p.join(p.dirname(_currentMigration.path), path));
```

Note that `p.join()` is a no-op if the final argument is already absolute.

You shouldn't need to set the load path at all in here; or rather, you should probably eventually have a `--load-path` CLI argument to match the executable's, but until then you can just assume there's no load path and all imports are relative.